cryptography/src/bls12381/primitives/ops/batch.rs

```diff
@@ -0,0 +1,262 @@
+//! was valid.
+
+use super::{
+    super::{group::Scalar, variant::Variant, Error},
+    hash_message, hash_message_namespace,
+};
+#[cfg(not(feature = "std"))]
+use alloc::{vec, vec::Vec};
+use commonware_math::algebra::{Random, Space};
+use rand_core::CryptoRngCore;
+
+/// Verifies multiple signatures over the same message from different public keys,
+/// ensuring each individual signature is valid.
+///
+/// Returns the indices of any invalid signatures found. Uses bisection internally
+/// to efficiently identify which signatures are invalid.
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on each public key
+/// and signature.
+pub fn verify_multiple_public_keys<R, V>(
+    rng: &mut R,
+    namespace: Option<&[u8]>,
+    message: &[u8],
+    entries: &[(V::Public, V::Signature)],
+) -> Vec<usize>
+where
+    R: CryptoRngCore,
+    V: Variant,
+{
+    if entries.is_empty() {
+        return Vec::new();
+    }
+
+    // Hash the message once
+    let hm = namespace.map_or_else(
+        || hash_message::<V>(V::MESSAGE, message),
+        |ns| hash_message_namespace::<V>(V::MESSAGE, ns, message),
+    );
+
+    // Iteratively bisect to find invalid signatures
+    let mut invalid = Vec::new();
+    let mut stack = vec![(0, entries.len())];
+    while let Some((start, end)) = stack.pop() {
+        let slice = &entries[start..end];
+        if slice.is_empty() {
+            continue;
+        }
+
+        // Generate random scalars for each signature in this slice
+        let scalars: Vec<Scalar> = (0..slice.len())
+            .map(|_| Scalar::random(&mut *rng))
+            .collect();
+
+        // Compute weighted sums: sum(r_i * pk_i) and sum(r_i * sig_i)
+        let pks: Vec<V::Public> = slice.iter().map(|(pk, _)| *pk).collect();
+        let sigs: Vec<V::Signature> = slice.iter().map(|(_, sig)| *sig).collect();
+        let weighted_pk = V::Public::msm(&pks, &scalars, 1);
+        let weighted_sig = V::Signature::msm(&sigs, &scalars, 1);
+
+        // Verify: e(weighted_pk, H(m)) == e(weighted_sig, G)
+        if V::verify(&weighted_pk, &hm, &weighted_sig).is_err() {
+            if slice.len() == 1 {
+                invalid.push(start);
+            } else {
+                let mid = slice.len() / 2;
+                stack.push((start + mid, end));
+                stack.push((start, start + mid));
+            }
+        }
+    }
+
+    invalid
+}
+
+/// Verifies multiple signatures over multiple unique messages from a single public key,
+/// ensuring each individual signature is valid.
+///
+/// Each entry is a tuple of (namespace, message, signature).
+///
+/// # Warning
+///
+/// This function assumes a group check was already performed on `public` and each `signature`.
+/// It is not safe to provide an aggregate public key or to provide duplicate messages.
+pub fn verify_multiple_messages<'a, R, V, I>(
```

Comment:
It actually **is** safe to provide duplicate messages, because of the random combinations, and we should update the documentation to reflect that, along with an explanation demonstrating that we considered this.